pkg/ottl/ottlfuncs/func_keep_matching_keys.go

```diff
@@ -0,0 +1,49 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+import (
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"golang.org/x/net/context"
+	"regexp"
+)
+
+type KeepMatchingKeysArguments[K any] struct {
+	Target  ottl.PMapGetter[K]
+	Pattern string
+}
+
+func NewKeepMatchingKeysFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("keep_matching_keys", &KeepMatchingKeysArguments[K]{}, createKeepMatchingKeysFunction[K])
+}
+
+func createKeepMatchingKeysFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*KeepMatchingKeysArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("KeepMatchingKeysFactory args must be of type *KeepMatchingKeysArguments[K")
+	}
+
+	return keepMatchingKeys(args.Target, args.Pattern)
+}
+
+func keepMatchingKeys[K any](target ottl.PMapGetter[K], pattern string) (ottl.ExprFunc[K], error) {
+	compiledPattern, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("the regex pattern provided to keep_matching_keys is not a valid pattern: %w", err)
+	}
```

Comment:
Could you move this logic to `createKeepMatchingKeysFunction`? Since the string is known at parse time, we should validate this on startup.

Reply:
@evan-bradley this is happening at startup since it is before the returned function.  We do all of our validations in this section across the ottlfuncs functions

Reply:
Whoops, good catch. I forgot we decided to do those here, thanks.